File: meetup13/phoenix_registration_and_channels/web/static/js/bank_account_channel.js
import {Socket} from "deps/phoenix/web/static/js/phoenix"


let server2client  = function(channel, bankAccount){
  channel.on("msg", (payload) => {
    bankAccount.html(payload.body)
  })
}


let setup = function(socket, user_id){

  let bankAccount = $("#bank-account");

  console.log("!!!");
  if (bankAccount.length > 0){
    console.log("2323");
    let channel = socket.channel("bank_account:" + user_id, {})

    channel.join()
      .receive("ok", resp => { console.log("Joined succesffuly to channel Bank Account", resp) })
      .receive("error", resp => { console.log("Unabled to join Bank Account", resp) })

    server2client(channel, bankAccount);

  }
}

export var BankAccountChannel = {setup: setup}

File: meetup13/phoenix_registration_and_channels/web/channels/bank_account_channel.ex
defmodule Swotter.BankAccountChannel do

  use Phoenix.Channel

  @msg "msg"

  def join("bank_account:" <> channel_id, auth_msg, socket) do

    IO.inspect "channel_id: #{channel_id}"

    user_id = Swotter.UserSocket.user_id(socket)

    channel_id = String.to_integer(channel_id)

    if user_id == channel_id do
      IO.inspect "user #{user_id} connected to channel BankAccount"
      {:ok, socket}
    else
      IO.inspect "channel subtopic #{channel_id} is different from user_id in socket #{user_id}"
      :error
    end
  end

  def send_to_socket(message, socket) do
    push socket, @msg, make_message(message)
  end

  def send_to_all_windows_of_user(message, user_id) do
    Swotter.Endpoint.broadcast! make_topic(user_id), @msg, make_message(message)
  end

  defp make_topic(user_id), do: "bank_account:#{user_id}"

  defp make_message(message), do: %{body: message}

end

# Swotter.BankAccountChannel.send_to_all_windows_of_user("hello there", 2)
